### ISSUE TITLE
Electron wakes up frequently even when running an app that is at rest

### ISSUE THREAD

**Issue by ckerr** [electron/electron#25021]
A downstream app is reporting that newer versions of Electron are causing more wakeups / CPU use due to IPC heartbeats. These don't appear to be specific to any particular app -- here are screenshots of profiling runs of the default app running in Electron Fiddle under different Electron versions:

Here are screenshots of profile runs of the default Electron app run in fiddle, with each version run multiple times for one minute. The results below represent roughly the average runtime requirements as measured by Windows Performance Recorder. 

* Electron 7.3.2: ![9c6fb4d5-09da-48ee-afa1-da99bc1791ef](https://user-images.githubusercontent.com/70381/90539089-0587cb80-e145-11ea-8b02-94905c1d4927.png) (6  + 89 + 4)

* Electron 8.4.1: ![dec6cba0-b474-4f77-ae0d-295b5e42fe4f](https://user-images.githubusercontent.com/70381/90539158-19cbc880-e145-11ea-81ea-074c9743fa2e.png) (34 + 164 + 3 + 49)

* Electron 9.1.0: ![0444902b-95fd-4f55-9115-f7e077c4df65](https://user-images.githubusercontent.com/70381/90539235-32d47980-e145-11ea-958e-6c1b6f9babb1.png) (140 + 4 + 2 + 19)

* Electron 9.1.1 (public vanilla version): ![6cb4b889-e014-4066-848b-085c73444d41](https://user-images.githubusercontent.com/70381/90539348-526ba200-e145-11ea-8c7f-f95a5380d887.png) (1 + 112 + 5)

It may be important that the first three versions are with a version of Electron that has been patched by downstream. I'm filing this bug for visibility / trackability and IMO the next step is to run similar tests on stock versions of Electron to see how much of this is reproducible in `electron/electron`.

### Preflight Checklist

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:**
  * 8.4.1
* **Operating System:**
  * Windows 10
* **Last Known Working Electron version:**
  * 7.3.2

### Expected Behavior
Electron apps that migrate to new versions of Electron should not experience more CPU use unless the app's code itself changes.

### Actual Behavior
"Hello World" app appears to use more cycles post-Electron 7

### To Reproduce
Profile an Electron app across multiple versions of Electron

**Comment by MarshallOfSound** [comment]
> At idle, Electron 8 seems to use more cycles than 7 or 9

Am I reading this right that this is _fixed_ in Electron 9?

**Comment by ckerr** [comment]
>> Profile an Electron app across multiple versions of Electron
>
> What profiling tool are you using to produce this info?

I did not do the profiling above, but I believe Windows Performance Recorder was used.

**Comment by ckerr** [comment]
>> At idle, Electron 8 seems to use more cycles than 7 or 9
>
> Am I reading this right that this is fixed in Electron 9?

It seems to be much better in 9, but not as good as 7.

*In the profiles above*:
 * 7 is by far the best
 * 8 is by far the worst
 * 9 is marginally worse than 7

But the bigger picture is -- even if 9 regresses compared to 7 -- I don't think we have anything actionable yet until we've done something like this with stock versions of Electron and perhaps longer runs in case there is statistical noise.

**Comment by ckerr** [comment]
OK I went in to try & reproduce this using Windows Performance Recorder and Windows Performance Analyzer on stock builds of Electron -- and found that these tools are very flexible, and record a wonderful amount of data; however, as I haven't used them before there is some learning curve involved. @zarubond could you share a recipe for testing on Windows?

I did a similar test on Linux by modding the default app to exit after 2 minutes, and then running electron in a perf stat loop, e.g.

```sh$ perf stat -r 5 /home/charles/electron/gn/8/src/out/testing/electron
```

The results are loosely similar:

```
 Performance counter stats for '/home/charles/electron/gn/7/src/out/testing/electron' (5 runs):

          2,812.66 msec task-clock                #    0.023 CPUs utilized            ( +-  1.98% )
             6,790      context-switches          #    0.002 M/sec                    ( +-  4.51% )
             1,148      cpu-migrations            #    0.408 K/sec                    ( +-  4.52% )
            74,744      page-faults               #    0.027 M/sec                    ( +-  0.24% )
     7,654,442,457      cycles                    #    2.721 GHz                      ( +-  1.70% )
     8,831,661,744      instructions              #    1.15  insn per cycle           ( +-  0.54% )
     1,610,531,197      branches                  #  572.600 M/sec                    ( +-  0.62% )
        34,073,211      branch-misses             #    2.12% of all branches          ( +-  1.59% )

          121.0650 +- 0.0229 seconds time elapsed  ( +-  0.02% )
```

```
 Performance counter stats for '/home/charles/electron/gn/8/src/out/testing/electron' (5 runs):

          3,230.75 msec task-clock                #    0.026 CPUs utilized            ( +-  1.44% )
             6,829      context-switches          #    0.002 M/sec                    ( +-  5.78% )
               292      cpu-migrations            #    0.090 K/sec                    ( +-  6.89% )
            73,426      page-faults               #    0.023 M/sec                    ( +-  0.27% )
     9,178,147,000      cycles                    #    2.841 GHz                      ( +-  1.43% )
     8,696,702,926      instructions              #    0.95  insn per cycle           ( +-  0.72% )
     1,578,664,989      branches                  #  488.638 M/sec                    ( +-  0.85% )
        33,251,860      branch-misses             #    2.11% of all branches          ( +-  1.32% )

          122.3774 +- 0.0522 seconds time elapsed  ( +-  0.04% )
```

```
 Performance counter stats for '/home/charles/electron/gn/9/src/out/testing/electron' (5 runs):

          1,442.90 msec task-clock                #    0.012 CPUs utilized            ( +-  3.41% )
             2,948      context-switches          #    0.002 M/sec                    ( +-  2.65% )
               512      cpu-migrations            #    0.355 K/sec                    ( +-  3.83% )
            43,895      page-faults               #    0.030 M/sec                    ( +-  0.19% )
     4,086,664,726      cycles                    #    2.832 GHz                      ( +-  1.26% )
     5,146,024,383      instructions              #    1.26  insn per cycle           ( +-  0.37% )
       923,994,850      branches                  #  640.373 M/sec                    ( +-  0.44% )
        17,502,342      branch-misses             #    1.89% of all branches          ( +-  0.72% )

          120.7651 +- 0.0182 seconds time elapsed  ( +-  0.02% )
```

**Comment by ckerr** [comment]
Same setup of the default app running for 120 seconds and then exiting. 10 runs against 10-x-y:

```
 Performance counter stats for '/home/charles/electron/gn/scratch/src/out/testing/electron' (10 runs):

          1,351.25 msec task-clock                #    0.011 CPUs utilized            ( +-  2.02% )
             2,630      context-switches          #    0.002 M/sec                    ( +-  2.22% )
               482      cpu-migrations            #    0.357 K/sec                    ( +-  6.20% )
            44,672      page-faults               #    0.033 M/sec                    ( +-  0.15% )
     3,631,612,713      cycles                    #    2.688 GHz                      ( +-  1.51% )
     4,413,432,036      instructions              #    1.22  insn per cycle           ( +-  0.45% )
       830,187,427      branches                  #  614.383 M/sec                    ( +-  0.50% )
        16,665,203      branch-misses             #    2.01% of all branches          ( +-  0.75% )

          120.7206 +- 0.0133 seconds time elapsed  ( +-  0.01% )
```

Slightly better than 9 even, even though 9 was so much better than 7 and 8. Pointer compression maybe?

If the original report is still an issue, it doesn't seem to be reproducible on Linux.

**Comment by zarubond** [comment]
I have been using Windows Performance Recorder which captures the measuring and saves it to file. You can many information including power, CPU, IO... The saved file is then opened with Windows Performance Analyzer to view them. On the left of this app you have a tab "Consumption" which gives you a list of processes and how mach impact they have on CPU. To get better idea what is happening you can also load symbols for electron through menu bar ->Trace -> Configure Symbol Paths. I have not able to make the online symbols provided by Electron work, so I was using our internal ones.  While the performance impact of Electron in idle is minimal, the issues is that it is waking up CPU from sleep which very inefficient.

Here is a run of Electron 9.2.1 over 10minutes, with basic Electron Fiddle example. You can see that the CPU is periodically woken up by IPC communication, but there isn`t happening any in user code.
![image](https://user-images.githubusercontent.com/6961047/91580259-1d353000-e94d-11ea-99e6-6ad874f0a70a.png)

During run I have also noticed a special process was forked to gather D3D info which was quite expensive.
![image](https://user-images.githubusercontent.com/6961047/91580551-92086a00-e94d-11ea-834d-01eb31c765e7.png)

**Comment by ckerr** [comment]
I did some more tracing on this today. While I don't have the cause tracked down yet (and so, no PR yet), I did want to check in with an update:

I did a walkthrough of places where DoWork is getting pumped in the message loop while the default app is idle, and also tried tracing out all the places where IPC messages are created while the default app is idle. It doesn't seem to be coming from any instances of `RepeatingTimer`s. However an `OnMemoryPressure()` call is being made every second or so, and that is causing mojo IPC messages to be created:

```cc
#0  0x0000555559466e14 in CreateMessage() () at ../../mojo/core/core.cc:341
#1  0x000055555c2c72d2 in CreateMessage () at ../../mojo/public/cpp/system/message.h:60
#2  CreateSerializedMessageObject () at ../../mojo/public/cpp/bindings/lib/message.cc:111
#3  Message() () at ../../mojo/public/cpp/bindings/lib/message.cc:246
#4  0x000055555c2c76e4 in mojo::Message::Message(unsigned int, unsigned int, unsigned long, unsigned long, std::__1::vector<mojo::ScopedHandleBase<mojo::Handle>, std::__1::allocator<mojo::ScopedHandleBase<mojo::Handle> > >*) () at ../../mojo/public/cpp/bindings/lib/message.cc:258
#5  0x000055555873af9f in OnMemoryPressure() () at gen/services/network/public/mojom/network_service.mojom.cc:2627
#6  0x000055555bef5716 in Run () at ../../base/callback.h:133
#7  Notify() () at ../../base/memory/memory_pressure_listener.cc:95
#8  0x000055555ae562b0 in Run () at ../../base/callback.h:133
#9  NotifyWrapper() () at ../../base/observer_list_threadsafe.h:248
#10 0x0000555558f4b805 in Invoke<void (media::GpuMemoryBufferVideoFramePool::PoolImpl::*)(media::GpuMemoryBufferVideoFramePool::PoolImpl::FrameResources*, gpu::SyncToken const&), scoped_refptr<media::GpuMemoryBufferVideoFramePool::PoolImpl>, media::GpuMemoryBufferVideoFramePool::PoolImpl::FrameResources*, gpu::SyncToken> () at ../../base/bind_internal.h:498
#11 MakeItSo<void (media::GpuMemoryBufferVideoFramePool::PoolImpl::*)(media::GpuMemoryBufferVideoFramePool::PoolImpl::FrameResources*, gpu::SyncToken const&), scoped_refptr<media::GpuMemoryBufferVideoFramePool::PoolImpl>, media::GpuMemoryBufferVideoFramePool::PoolImpl::FrameResources*, gpu::SyncToken> () at ../../base/bind_internal.h:637
#12 RunImpl<void (media::GpuMemoryBufferVideoFramePool::PoolImpl::*)(media::GpuMemoryBufferVideoFramePool::PoolImpl::FrameResources*, gpu::SyncToken const&), std::__1::tuple<scoped_refptr<media::GpuMemoryBufferVideoFramePool::PoolImpl>, media::GpuMemoryBufferVideoFramePool::PoolImpl::FrameResources*, gpu::SyncToken>, 0, 1, 2> () at ../../base/bind_internal.h:710
#13 RunOnce() () at ../../base/bind_internal.h:679
#14 0x000055555bf38f79 in Run () at ../../base/callback.h:99
#15 RunTask() () at ../../base/task/common/task_annotator.cc:165
```
When I pick up on this on Monday, my next step is to see if that is indeed causing the every-second wakeups that it appears to be doing and what the source of that event is.

**Comment by ckerr** [comment]
I will have more time this week to look into this and have also pinged @zcbenz to see if he has time as well.

**Comment by ckerr** [comment]
Followup to the previous lead: unfortunately, the memory pressure [is coming from](https://github.com/electron/electron/blob/master/shell/browser/electron_browser_main_parts.cc#L416) a GC timer that we fire every minute:

```c++
  // Start idle gc.
  gc_timer_.Start(
      FROM_HERE, base::TimeDelta::FromMinutes(1),
      base::Bind(&v8::Isolate::LowMemoryNotification,
                 base::Unretained(js_env_->isolate())));
```
This has been in Electron [since 2014](https://github.com/electron/electron/commit/4dd78480846557370dd53789d50e6cb0d6a81bcd) so it's unlikely to be the cause of extra wakeups in newer versions :slightly_smiling_face: but it may be worth looking at as a side-issue to see if this could have a resource-based trigger rather than a time-based one.

**Comment by zcbenz** [comment]
I can not observe obvious difference between 7 and 8's CPU cycles.

I'm running the recorder under a Windows 10 virtual machine, measuring the default app of Electron. The default app was measured a few seconds after it got started, it was put in background and the recorder window was focused during recording.

Here are the results of 7.3.3:
<img width="696" alt="7-3-3-1" src="https://user-images.githubusercontent.com/639601/96080136-663b4680-0ef1-11eb-8a4d-4b066fa6ef67.png">
<img width="580" alt="7-3-3-2" src="https://user-images.githubusercontent.com/639601/96080140-676c7380-0ef1-11eb-8dac-df61db327c8c.png">
<img width="579" alt="7-3-3-3" src="https://user-images.githubusercontent.com/639601/96080145-69363700-0ef1-11eb-9ca5-a1ccd4bb730e.png">

Here are the results of 8.5.2:
<img width="696" alt="8-5-2-1" src="https://user-images.githubusercontent.com/639601/96080171-705d4500-0ef1-11eb-9fd6-f85980c4c0e7.png">
<img width="575" alt="8-5-2-2" src="https://user-images.githubusercontent.com/639601/96080173-718e7200-0ef1-11eb-84b2-54c69222bb50.png">
<img width="575" alt="8-5-2-3" src="https://user-images.githubusercontent.com/639601/96080454-0e510f80-0ef2-11eb-9257-81c2364afb87.png">

**Comment by ckerr** [comment]
The once-per-minute wakeup has landed in master and is targeted for backport to 11 and 10.

**Comment by ckerr** [comment]
The backtraces obtainable in a debugger from the DoWork loop are not particularly helpful, so I tried a different way of measuring traffic:

* I instrumented the base libraries to log information when a new task was created to be run through the event loop. I used the "posted_from" field in `base::PostedTask` to log the _source_ of the task since this generally gives the root source, whereas a stacktrace mostly shows that tasks are generated in an event loop callback that's layers of indirection removed from the source.
* I ran the default app overnight with this instrumentation.
* I clipped off the first few minutes of the log to ensure only "steady state" traffic was measured
* I sorted by frequency and took the top 50 line items.

The results are below

### Next Steps

The memory pressure monitor generated 10x more tasks than any other source. My next step will be to investigate if all of those tasks are necessary.

### Top 50 task sources (steady-state only)

|Calls|Posted From|
|--:|:--|
|10093|StartObserving@../../base/util/memory_pressure/system_memory_pressure_evaluator_win.cc:238|
|10093|StartMetricsTimer@../../base/util/memory_pressure/multi_source_memory_pressure_monitor.cc:45|
|801|Notify@../../mojo/public/cpp/system/simple_watcher.cc:100|
|360|DidHandleInputEventSentToMain@../../third_party/blink/renderer/platform/widget/input/widget_input_handler_manager.cc:797|
|285|PostDispatchNextMessageFromPipe@../../mojo/public/cpp/bindings/lib/connector.cc:530|
|252|HighResolutionTimerManager@../../base/timer/hi_res_timer_manager_win.cc:50|
|168|StartTrimRulesTask@../../extensions/browser/api/declarative_net_request/action_tracker.cc:443|
|167|GpuDataManagerImplPrivate@../../content/browser/gpu/gpu_data_manager_impl_private.cc:494|
|166|ReportMetrics@../../third_party/blink/renderer/core/execution_context/agent_metrics_collector.cc:94|
|165|StartMonitoringIfNeeded@../../third_party/blink/renderer/controller/memory_usage_monitor.cc:46|
|62|PostNextTickTask@../../components/viz/common/frame_sinks/delay_based_time_source.cc:159|
|59|ScheduleBeginImplFrameDeadline@../../cc/scheduler/scheduler.cc:716|
|39|ScheduledActionSendBeginMainFrame@../../cc/trees/proxy_impl.cc:605|
|38|SendCommitRequestToImplThreadIfNeeded@../../cc/trees/proxy_main.cc:669|
|37|PerformHeartbeatLatencyReport@../../base/task/thread_pool/service_thread.cc:83|
|37|BeginMainFrame@../../cc/trees/proxy_main.cc:333|
|36|Init@../../base/task/thread_pool/service_thread.cc:47|
|32|SetNeedsMainFrame@../../third_party/blink/renderer/platform/widget/input/main_thread_event_queue.cc:692|
|28|SpellCheckHostMetrics@../../components/spellcheck/browser/spellcheck_host_metrics.cc:27|
|27|TaskSetFinished@../../cc/tiles/tile_manager.cc:329|
|26|WorkNow@../../net/dns/serial_worker.cc:34|
|21|Schedule@../../cc/base/unique_notifier.cc:35|
|19|StartWatching@../../base/win/registry.cc:89|
|15|SendMessageW@../../ipc/ipc_mojo_bootstrap.cc:731|
|15|Accept@../../ipc/ipc_mojo_bootstrap.cc:897|
|14|ScheduleBeginFrameDeadline@../../components/viz/service/display/display_scheduler.cc:345|
|14|RunOnWorkerThread@../../cc/tiles/tile_manager.cc:353|
|13|ScheduleProcessRipeTasksOnServiceThread@../../base/task/thread_pool/delayed_task_manager.cc:155|
|13|PushOntoDelayedIncomingQueue@../../base/task/sequence_manager/task_queue_impl.cc:430|
|11|DidNotSwap@../../third_party/blink/renderer/core/frame/web_frame_widget_base.cc:1398|
|8|ProcessEventHookCallback@../../ui/aura/native_window_occlusion_tracker_win.cc:649|
|8|PostTaskToMainThread@../../third_party/blink/renderer/platform/widget/input/main_thread_event_queue.cc:554|
|7|Notify@../../mojo/public/cpp/system/simple_watcher.cc:100|
|6|UpdateIfNeeded@../../ui/display/win/color_profile_reader.cc:73|
|6|StartTimer@../../net/dns/dns_config_service.cc:136|
|6|ScheduleOcclusionCalculationIfNeeded@../../ui/aura/native_window_occlusion_tracker_win.cc:481|
|6|ScheduledActionSendBeginMainFrame@../../cc/trees/single_thread_proxy.cc:765|
|6|OnSwapBuffers@../../components/viz/service/display/software_output_device.cc:57|
|6|DidReceiveCompositorFrameAckOnImplThread@../../cc/trees/single_thread_proxy.cc:491|
|6|DidPresentCompositorFrame@../../cc/trees/layer_tree_host_impl.cc:2017|
|6|ComputeNativeWindowOcclusionStatus@../../ui/aura/native_window_occlusion_tracker_win.cc:472|
|5|OnObjectSignaled@../../base/files/file_path_watcher_win.cc:195|
|5|OnBeginFrame@../../components/viz/service/display/display_scheduler.cc:159|
|5|Notify@../../mojo/public/cpp/system/simple_watcher.cc:1002020-10-26T23:49:17.292Z|
|5|EmitEventSoon@../../electron\shell/browser/api/electron_api_base_window.h:250|
|4|Start@../../base/allocator/partition_allocator/memory_reclaimer.cc:97|
|4|PostPendingBeginFrameTask@../../cc/scheduler/scheduler.cc:314|
|3|RequestHDRStatus@../../content/browser/gpu/gpu_data_manager_impl_private.cc:435|
|3|RequestHDRStatus@../../components/viz/service/gl/gpu_service_impl.cc:774|

### The top 50 task sources (startup included)

|Calls|Posted From|
|--:|:--|
|10102|StartObserving@../../base/util/memory_pressure/system_memory_pressure_evaluator_win.cc:238|
|10102|StartMetricsTimer@../../base/util/memory_pressure/multi_source_memory_pressure_monitor.cc:45|
|1585|Notify@../../mojo/public/cpp/system/simple_watcher.cc:100|
|360|DidHandleInputEventSentToMain@../../third_party/blink/renderer/platform/widget/input/widget_input_handler_manager.cc:797|
|343|PostDispatchNextMessageFromPipe@../../mojo/public/cpp/bindings/lib/connector.cc:530|
|254|HighResolutionTimerManager@../../base/timer/hi_res_timer_manager_win.cc:50|
|170|StartMonitoringIfNeeded@../../third_party/blink/renderer/controller/memory_usage_monitor.cc:46|
|169|StartTrimRulesTask@../../extensions/browser/api/declarative_net_request/action_tracker.cc:443|
|168|GpuDataManagerImplPrivate@../../content/browser/gpu/gpu_data_manager_impl_private.cc:494|
|167|ReportMetrics@../../third_party/blink/renderer/core/execution_context/agent_metrics_collector.cc:94|
|102|SendMessageW@../../ipc/ipc_mojo_bootstrap.cc:731|
|100|Accept@../../ipc/ipc_mojo_bootstrap.cc:897|
|78|PostNextTickTask@../../components/viz/common/frame_sinks/delay_based_time_source.cc:159|
|74|ScheduleBeginImplFrameDeadline@../../cc/scheduler/scheduler.cc:716|
|65|ArmOrNotify@../../mojo/public/cpp/system/simple_watcher.cc:234|
|49|TaskSetFinished@../../cc/tiles/tile_manager.cc:329|
|42|ScheduledActionSendBeginMainFrame@../../cc/trees/proxy_impl.cc:605|
|42|Schedule@../../cc/base/unique_notifier.cc:35|
|41|SendCommitRequestToImplThreadIfNeeded@../../cc/trees/proxy_main.cc:669|
|39|Init@../../base/task/thread_pool/service_thread.cc:47|
|37|PerformHeartbeatLatencyReport@../../base/task/thread_pool/service_thread.cc:83|
|37|BeginMainFrame@../../cc/trees/proxy_main.cc:333|
|36|WakeupMainThread@../../electron/shell/common/node_bindings.cc:553|
|34|WorkNow@../../net/dns/serial_worker.cc:34|
|32|SetNeedsMainFrame@../../third_party/blink/renderer/platform/widget/input/main_thread_event_queue.cc:692|
|29|SpellCheckHostMetrics@../../components/spellcheck/browser/spellcheck_host_metrics.cc:27|
|27|Accept@../../mojo/public/cpp/bindings/thread_safe_forwarder_base.cc:56|
|25|RunOnWorkerThread@../../cc/tiles/tile_manager.cc:353|
|22|StartWatching@../../base/win/registry.cc:89|
|22|ScheduleBeginFrameDeadline@../../components/viz/service/display/display_scheduler.cc:345|
|19|PushOntoDelayedIncomingQueue@../../base/task/sequence_manager/task_queue_impl.cc:430|
|17|ScheduleProcessRipeTasksOnServiceThread@../../base/task/thread_pool/delayed_task_manager.cc:155|
|12|AddFilter@../../ipc/ipc_channel_proxy.cc:303|
|11|Start@../../mojo/core/channel_win.cc:107|
|11|Notify@../../mojo/public/cpp/system/simple_watcher.cc:100|
|11|DidPresentCompositorFrame@../../cc/trees/layer_tree_host_impl.cc:2017|
|11|DidNotSwap@../../third_party/blink/renderer/core/frame/web_frame_widget_base.cc:1398|
|11|BindInterface@../..\services/service_manager/public/cpp/interface_binder.h:60|
|10|BeginRead@../../third_party/blink/renderer/platform/loader/fetch/response_body_loader.cc:52|
|9|UpdateIfNeeded@../../ui/display/win/color_profile_reader.cc:73|
|9|ScheduledActionSendBeginMainFrame@../../cc/trees/single_thread_proxy.cc:765|
|9|ProcessEventHookCallback@../../ui/aura/native_window_occlusion_tracker_win.cc:649|
|9|PostTaskToMainThread@../../third_party/blink/renderer/platform/widget/input/main_thread_event_queue.cc:554|
|9|OnSwapBuffers@../../components/viz/service/display/software_output_device.cc:57|
|8|Send@../../ipc/ipc_channel_proxy.cc:428|
|8|ScheduleOcclusionCalculationIfNeeded@../../ui/aura/native_window_occlusion_tracker_win.cc:481|
|8|DidReceiveCompositorFrameAckOnImplThread@../../cc/trees/single_thread_proxy.cc:491|
|8|CreateAsarURLLoader@../../electron/shell/browser/net/asar/asar_url_loader.cc:314|
|8|ComputeNativeWindowOcclusionStatus@../../ui/aura/native_window_occlusion_tracker_win.cc:472|

**Comment by ckerr** [comment]
The system memory pressure evaluator is a periodic checkup to see what level of memory is being used.

Its repeat interval is defined in `base/memory/memory_pressure_monitor.cc` to be triggered every five seconds:

```c++
  const base::TimeDelta MemoryPressureMonitor::kUMAMemoryPressureLevelPeriod =
      base::TimeDelta::FromSeconds(5);
```

This periodic check exists on both Windows and Linux, but on Macs we plug into [`DISPATCH_SOURCE_TYPE_MEMORYPRESSURE`](https://developer.apple.com/documentation/dispatch/dispatch_source_type_memorypressure) to let the OS notify us when the memory pressure changes, so Macs _don't_ see the same wakeup.

It looks like Chromium has some support for subscribing to OS signals iff on Win 8.1 or higher. This appears to be experimental code that was [added to Chromium in Aug 2020](https://chromium-review.googlesource.com/c/chromium/src/+/2321524). The code is disabled behind a feature flag _and_ when updated, augments the periodic timer rather than replacing it:

> This adds a second voter to the Win system memory pressure evaluator that will base its vote on the OS Memory Pressure signals. This voter works in parallel to the default voter (based on heuristics) and so the pressure level will always be based on worst of these 2 signals.
>
> This code is behind a feature flag, some feature param will later be added to test some variations of this solution (e.g. only use the OS-signals based voter for the critical pressure notifications).

If/when this migration to only using OS-signals lands, then the every-5-seconds wakeup will go away on Windows.

**Comment by ckerr** [comment]
Also, note that this wakeup happens 10x more often than any other.

**Comment by ckerr** [comment]
On the Linux side there doesn't seem to be a clean solution that's already shipping; however, https://gitlab.freedesktop.org/hadess/low-memory-monitor/ might be one possible option if it ever moves into production:

> ... when memory pressure means that memory isn't as readily available and would cause interactivity problems, would: send D-Bus signals to user-space applications when memory is running low

**Comment by pushkin-** [comment]
@ckerr Should this be reopened? Do you know if this is still an issue?
